### ISSUE TITLE
Uninstantiated generic types as generic type arguments

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#12027]
Normally, all generic type arguments of a generic cannot themselves be uninstantiated generics:

```crystal
Comparable(Array)            # Error: can't use Array(T) as generic type argument yet, use a more specific type
[1].is_a?(Comparable(Array)) # Error: can't use Array(T) as generic type argument yet, use a more specific type
```

But such generics can be formed nonetheless:

```crystal
class Foo(T); end
module Moo(T); end

alias F = Foo(Foo) # okay
alias M = Moo(Moo) # okay

class Bar < F      # okay
  include Moo(Moo) # okay
end

{% Bar.ancestors %} # => [Moo(Moo(T)), Foo(Foo(T)), Reference, Object]

Bar.new.is_a?(F) # => true
Bar.new.is_a?(M) # => true
```

Formal type parameters can be leaked in this way too:

```crystal
class Foo(T)
  @x : T?
end

module Moo(T)
  @y : T?
end

typeof(Bar.new.@x) # => (Foo(T) | Nil)
typeof(Bar.new.@y) # => (Moo(T) | Nil)
```

So far we have been using types like `Comparable(Array)` in the standard library without any significant issues. What should we do about these types?

**Comment by beta-ziliani** [comment]
Perhaps obvious, but it's important to mention that `Bar.new.@x` makes the program crush.

```
cast from Crystal::GenericModuleType to Crystal::SubclassObservable failed, at /build/crystal/src/crystal-1.4.1/src/compiler/crystal/semantic/call.cr:1229:5:1229 (TypeCastError)
```

**Comment by asterite** [comment]
I don't think we should let un instantiated generic types as instance variables. It's allowed for include to use T as a generic type restriction (Comparable is a good example of that).

Regarding ancestors... I guess uninstantiated generic types make sense In a hierarchy.

**Comment by straight-shoota** [comment]
The argument type for `Array#<=>` is not really an unrestricted `Array` type. It's actually something like `Array(Comparable(T))`.
That's technically not entirely accurate, because the comparison is the other way around and it would be more like `Array(Comparable(U))` for all U such that `T < Comparable(U)`.
That's just too complex to express.

I believe we could probably reverse the call direction of the comparison in `Array<=>` so that we can make the argument an actual `Array(Comparable(T))`.

Implementing this triggers a compiler bug and sends it into an endless loop (https://github.com/straight-shoota/crystal/tree/test/array-comparable).

**Comment by asterite** [comment]
> The argument type for Array#<=> is not really an unrestricted Array type

It is! Take a look at this:

```crystal
class Foo
  getter x

  def initialize(@x : Int32)
  end

  def <=>(other : Foo)
    x <=> other.x
  end
end

ary = [Foo.new(3), Foo.new(2), Foo.new(1)]
p ary.sort
```

It just works ©️ 

> That's just too complex to express

This is something that I really like about Crystal: no need to delve into super complex type restrictions.

If we do restrict it to `Array(Comparable(T))` then it's a breaking change. But don't feel it's the right thing to do. In my mind `Comparable` just defines `<` and others for you based on a single `<=>` method, but it doesn't signal that a type is comparable. Similar to how a type can respond to `to_json` or `from_json` without necessarily including `JSON::Serializable`. Whether that's something desireable or not is a different topic, but it's how things work (more like a duck-type way, which is what I really like about Ruby.)

**Comment by straight-shoota** [comment]
@asterite I don't get your code example. We're talking about `Array#<=>`, not `#sort` (which is eventually `T#<=>`).

I sympathize with the ability to avoid complexity.
But there is a good motivation to adding such type restrictions: Better errors.
Right now, if you do something like `[1] <=> ['c']` you get an error message that complains about undefined `Int32#<=>(String)`. That's not very helpful. It might be good enough to figure out what's wrong.
But it would be more user friendly if `Array#<=>` wouldn't even compile when the item types are not comparable.

**Comment by asterite** [comment]
Ah, you are right. Here's a better example:

```crystal
class Foo
  getter x

  def initialize(@x : Int32)
  end

  def <=>(other : Foo)
    x <=> other.x
  end
end

ary1 = [Foo.new(3), Foo.new(2), Foo.new(1)]
ary2 = [Foo.new(1), Foo.new(2), Foo.new(1)]
p ary1 <=> ary2
```

> Right now, if you do something like [1] <=> ['c'] you get an error message that complains about undefined Int32#<=>(String)

That makes total sense if you look at the entire error trace... which is hidden by default 🤦

**Comment by asterite** [comment]
> Right now, if you do something like [1] <=> ['c'] you get an error message that complains about undefined Int32#<=>(String)

A better answer that doesn't use the facepalm emoji (sorry about that): we could check if each element responds to `<=>` with the target type. If not, we can give a nice error message. But I don't think enforcing this at the type level is the best solution.
